### PR TITLE
Create `Preferences` model, enum `Role`

### LINKED ISSUE #84: Add model and viewmodel layer for WoofApp to handle app flow
## Depends on tasks:

- #79
- #110 

## Description
- [x] create model layer containing information about user preferences `Preferences` - #111 
- [ ] create ViewModel that will provide data or status to handle AppFlow (sitter or owner side).

#### `WoofApp.swift`
- create model named `Preferences` containing next fields
```swift
var selectedRole: Role?
```

```swift
enum Role{
    case owner, sitter
}
```
- creating a ViewModel in which the following should be implemented
- creating property to inform view(WoofApp) about loaded selected role
``` swift
@Published var loadRole: Role? 
```
- creating a read method to get the selected role from preferences KeyValueStorage
- creating tests for methods of ViewModel, if needed.

### PR DESCRIPTION
**Pull request by NadzeyaShpakouskaya**
## Summary
This PR is a part of the work to resolve #84.

## Changes
- Add the `Preferences` model. 
- Add `Role` enum.
- Add tests for new objects.

## Checks to complete
- [x] I've built and run my changes, and no warnings or errors occurred.
- [x] All existing tests passed with my changes.
- [x] My code follows our style guide.
- [x] I've performed a self-review of my code.
- [x] PR's title has a conventional commit style (short and descriptive).
- [x] Every non-private interface is documented properly.
- [x] I've added tests for my code (if it makes sense).
- [x] PR has assignee, reviewers, milestone, it is properly linked to the project and to the issue.

### CODE REVIEW

**Review by julyakmita — APPROVED** [review]
LGTM.

File: Woof/Woof/Model/Preferences/Preferences.swift [hunk 1]
```diff
@@ -0,0 +1,8 @@
+import Foundation
+
+/// Represents user preferences.
+struct Preferences: Codable {
+    /// The selected role for the user.
+    /// If role is not selected returns `nil`.
+    var selectedRole: Role?
```
Comment (dersim-davaod) [inline Woof/Woof/Model/Preferences/Preferences.swift:7]: I'd suggest to avoid optional here. Instead you can add a new `none` case in the enum.

File: Woof/Woof/Model/Preferences/Preferences.swift [hunk 2]
```diff
@@ -0,0 +1,8 @@
+import Foundation
+
+/// Represents user preferences.
+struct Preferences: Codable {
```
Comment (dersim-davaod) [inline Woof/Woof/Model/Preferences/Preferences.swift:4]: Do we need a conformance to `Codable` at the moment?

File: Woof/Woof/Model/Preferences/Role.swift [hunk 3]
```diff
@@ -0,0 +1,10 @@
+import Foundation
+
+/// Provides the roles that can be chosen in the app.
```
Comment (dersim-davaod) [inline Woof/Woof/Model/Preferences/Role.swift:3]: nit:
```suggestion
/// A role of a user in the app's session. A user can have one active role.
```

File: Woof/Woof/Model/Preferences/Role.swift [hunk 4]
```diff
@@ -0,0 +1,10 @@
...
+enum Role: Codable {
+    /// Represents the owner role.
```
Comment (dersim-davaod) [inline Woof/Woof/Model/Preferences/Role.swift:5]: nit:

```suggestion
    /// An owner of the pet.
```

Comment (dersim-davaod) [reply Woof/Woof/Model/Preferences/Role.swift:5]: Please review the rest of code.

**Review by dersim-davaod — CHANGES_REQUESTED** [review]
Consider improvements:

**Review by dersim-davaod — CHANGES_REQUESTED** [review]
Consider improvements:

File: Woof/Woof/Model/Preferences/Preferences.swift (hunk 1 above)
Comment (NadzeyaShpakouskaya) [reply Woof/Woof/Model/Preferences/Preferences.swift]: Agree, it helps us to avoid additional unwrapping in the future.

File: Woof/Woof/Model/Preferences/Preferences.swift (hunk 2 above)
Comment (NadzeyaShpakouskaya) [reply Woof/Woof/Model/Preferences/Preferences.swift:4]: We store Preferences in KeyValueStorage as Data, that's why we need to conform to the Codable Protocol.

File: Woof/Woof/Model/Preferences/Role.swift (hunk 3 above)
Comment (NadzeyaShpakouskaya) [reply Woof/Woof/Model/Preferences/Role.swift]: Fixed.

File: Woof/Woof/Model/Preferences/Role.swift (hunk 4 above)
Comment (NadzeyaShpakouskaya) [reply Woof/Woof/Model/Preferences/Role.swift]: Fixed.

File: Woof/WoofTests/Model/Preferences/RoleTests.swift [hunk 5]
```diff
@@ -0,0 +1,8 @@
+import XCTest
+
+final class RoleTests: XCTestCase {
+    func testAllRolesExistInAPI() {
+        _ = Role.owner
+        _ = Role.sitter
```
Comment (dersim-davaod) [inline Woof/WoofTests/Model/Preferences/RoleTests.swift:6]: Should we add the test case for `none`?
Should we use the [CaseIterable](https://developer.apple.com/documentation/swift/caseiterable) protocol and its `allCases` property to check the existence of all roles?

**Review by dersim-davaod — COMMENTED** [review]
Overall LGTM, but have one more question:

File: Woof/WoofTests/Model/Preferences/RoleTests.swift (hunk 5 above)
Comment (NadzeyaShpakouskaya) [reply Woof/WoofTests/Model/Preferences/RoleTests.swift]: Added the conformation to CaseIterable.

File: Woof/WoofTests/Model/Preferences/RoleTests.swift [hunk 6]
```diff
@@ -0,0 +1,9 @@
+import XCTest
+
+final class RoleTests: XCTestCase {
```
Comment (iBYarchitect) [inline Woof/WoofTests/Model/Preferences/RoleTests.swift:3]: Does it make sense to remake the test in such a way as to check cases or, in principle, does the original version check for the presence of cases?

```suggestion
 final class RoleTests: XCTestCase {
    func testAllRolesExistInAPI() {
        Role.allCases.forEach { role in
            XCTAssertTrue(checkRoleExistsInAPI(role))
        }
    }

    func checkRoleExistsInAPI(_ role: Role) -> Bool {
        switch role {
        case .owner,
             .sitter,
             .none:
            return true
        }
    }
 }
```

**Review by iBYarchitect — COMMENTED** [review]
There is one question regarding testing the availability of all cases:

File: Woof/WoofTests/Model/Preferences/RoleTests.swift (hunk 6 above)
Comment (NadzeyaShpakouskaya) [reply Woof/WoofTests/Model/Preferences/RoleTests.swift:3]: I agree, that my code does nothing 🤦‍♀️ 
Fixed with https://github.com/ios-course/ironfoudation-team-project/pull/111/commits/0970ea72227e9b020965214b63f454899ec4556c

**Review by iBYarchitect — APPROVED** [review]
LGTM

**Review by dersim-davaod — APPROVED** [review]
LGTM
